api4/config_local.go

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package api4
+
+import (
+	"net/http"
+	"reflect"
+
+	"github.com/mattermost/mattermost-server/v5/audit"
+	"github.com/mattermost/mattermost-server/v5/config"
+	"github.com/mattermost/mattermost-server/v5/model"
+	"github.com/mattermost/mattermost-server/v5/utils"
+)
+
+func (api *API) InitConfigLocal() {
+	api.BaseRoutes.ApiRoot.Handle("/config", api.ApiLocal(getConfig)).Methods("GET")
+	api.BaseRoutes.ApiRoot.Handle("/config", api.ApiLocal(localUpdateConfig)).Methods("PUT")
+	api.BaseRoutes.ApiRoot.Handle("/config/patch", api.ApiLocal(localPatchConfig)).Methods("PUT")
+}
+
+func localUpdateConfig(c *Context, w http.ResponseWriter, r *http.Request) {
+	cfg := model.ConfigFromJson(r.Body)
+	if cfg == nil {
+		c.SetInvalidParam("config")
+		return
+	}
+
+	auditRec := c.MakeAuditRecord("localUpdateConfig", audit.Fail)
+	defer c.LogAuditRec(auditRec)
+
+	cfg.SetDefaults()
+
+	appCfg := c.App.Config()
+
+	// Do not allow plugin uploads to be toggled through the API
+	cfg.PluginSettings.EnableUploads = appCfg.PluginSettings.EnableUploads
+
+	// Do not allow certificates to be changed through the API
+	cfg.PluginSettings.SignaturePublicKeyFiles = appCfg.PluginSettings.SignaturePublicKeyFiles
+
+	c.App.HandleMessageExportConfig(cfg, appCfg)
+
+	err := cfg.IsValid()
+	if err != nil {
+		c.Err = err
+		return
+	}
+
+	err = c.App.SaveConfig(cfg, true)
+	if err != nil {
+		c.Err = err
+		return
+	}
+
+	cfg = c.App.GetSanitizedConfig()
```

Comment:
Do we need to sanitize for local mode?

Reply:
The current idea is to keep the local endpoints working as the remote API ones would do for a sysadmin as much as possible, so this is just following that trend

Reply:
Fair enough 👍